### PR TITLE
KAA-433: Fix extension counting while serialization.

### PR DESCRIPTION
**Pull request by ohamov**

### CODE REVIEW

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 1]
```diff
@@ -342,9 +343,19 @@ static kaa_error_t kaa_client_sync_serialize(kaa_platform_protocol_t *self
         }
         case KAA_SERVICE_LOGGING: {
 #ifndef KAA_DISABLE_FEATURE_LOGGING
-            error_code = kaa_logging_request_serialize(self->kaa_context->log_collector, writer);
-            if (error_code)
-                KAA_LOG_ERROR(self->logger, error_code, "Failed to serialize the logging extension");
+            bool need_resync = false;
+            error_code = kaa_logging_need_logging_resync(self->kaa_context->log_collector, &need_resync);
+            if(!error_code) {
+                if(need_resync) {
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c]: Code style spaces.

File: client/client-multi/client-c/src/kaa/kaa_platform_protocol.c [hunk 2]
```diff
@@ -342,9 +343,19 @@ static kaa_error_t kaa_client_sync_serialize(kaa_platform_protocol_t *self
         }
         case KAA_SERVICE_LOGGING: {
 #ifndef KAA_DISABLE_FEATURE_LOGGING
-            error_code = kaa_logging_request_serialize(self->kaa_context->log_collector, writer);
-            if (error_code)
-                KAA_LOG_ERROR(self->logger, error_code, "Failed to serialize the logging extension");
+            bool need_resync = false;
+            error_code = kaa_logging_need_logging_resync(self->kaa_context->log_collector, &need_resync);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_platform_protocol.c:347]: I don't see the kaa_logging_need_logging_resync definition.

File: client/client-multi/client-c/test/test_protocol.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+extern kaa_error_t kaa_channel_manager_create(kaa_channel_manager_t **channel_manager_p, kaa_context_t *context);
+extern kaa_error_t kaa_log_collector_create(kaa_log_collector_t ** log_collector_p
+                                          , kaa_status_t *status
+                                          , kaa_channel_manager_t *channel_manager
+                                          , kaa_logger_t *logger);
+
+extern kaa_error_t kaa_platform_protocol_create(kaa_platform_protocol_t **platform_protocol_p, kaa_context_t *context
+                                              , kaa_status_t *status);
+
+extern kaa_error_t kaa_platform_protocol_serialize_client_sync(kaa_platform_protocol_t *self, const kaa_serialize_info_t *info
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: The same as above.

File: client/client-multi/client-c/test/test_protocol.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                                                             , char **buffer, size_t *buffer_size);
+
+extern kaa_error_t kaa_deinit(kaa_context_t *kaa_context);
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: The same as above.

File: client/client-multi/client-c/test/test_protocol.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+#include "kaa_configuration_manager.h"
+#include "kaa_status.h"
+#include "utilities/kaa_mem.h"
+#include "utilities/kaa_log.h"
+#include "platform/ext_log_storage.h"
+#include "platform/ext_log_upload_strategy.h"
+
+#ifndef KAA_DISABLE_FEATURE_LOGGING
+extern kaa_error_t kaa_status_create(kaa_status_t **kaa_status_p);
+extern kaa_error_t kaa_init(kaa_context_t **kaa_context_p);
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: This extern is not neeed as the declaration presents in the corresponding header.

File: client/client-multi/client-c/test/test_protocol.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+{
+    kaa_service_t service = KAA_SERVICE_LOGGING;
+    info = (kaa_serialize_info_t *) malloc(sizeof(kaa_serialize_info_t));
+    info->services = &service;
+    info->services_count = 1;
+    info->allocator = &allocator;
+    info->allocator_context = mock;
+    mock_strategy_context_t *strategy = (mock_strategy_context_t*) malloc(sizeof(mock_strategy_context_t));
+    mock_storage_context_t *storage = (mock_storage_context_t*) malloc(sizeof(mock_storage_context_t));
+    if(!storage || !strategy) {
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: Please, use ASSERT_NOT_NULL instead.

File: client/client-multi/client-c/test/test_protocol.c [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    size_t record_count;
+    size_t total_size;
+    bool on_remove_by_id_count;
+    bool on_unmark_by_id_count;
+} mock_storage_context_t;
+
+void test_empty_log_collector_extension_count(void)
+{
+    kaa_service_t service = KAA_SERVICE_LOGGING;
+    info = (kaa_serialize_info_t *) malloc(sizeof(kaa_serialize_info_t));
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: There is several memory leaks. Please, fix them.

File: client/client-multi/client-c/test/test_protocol.c [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    ASSERT_EQUAL(count_of_extensions, 1);
+}
+int initialize()
+{
+    kaa_error_t error_code = kaa_init(&kaa_context);
+    if(error_code) {
+    	exit(error_code);
+    }
+    error_code = kaa_status_create(&status);
+    if(error_code) {
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: Code style spaces.

File: client/client-multi/client-c/test/test_protocol.c [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    ASSERT_EQUAL(error_code, KAA_ERR_NONE);
+    char count_of_extensions = *(buffer + 7);
+    KAA_LOG_DEBUG(kaa_context->logger, KAA_ERR_NONE, "count of extensions is %d", count_of_extensions);
+    ASSERT_EQUAL(count_of_extensions, 1);
+}
+int initialize()
+{
+    kaa_error_t error_code = kaa_init(&kaa_context);
+    if(error_code) {
+    	exit(error_code);
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: Please, use "return error_code" instead.

File: client/client-multi/client-c/test/test_protocol.c [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    }
+    error_code = kaa_channel_manager_create(&channel_manager, kaa_context);
+    if(error_code) {
+        exit(error_code);
+        }
+    error_code = kaa_log_collector_create(&log_collector, status, channel_manager, logger);
+    if(error_code) {
+        exit(error_code);
+        }
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: Please, align braces all around the functions.

File: client/client-multi/client-c/test/test_protocol.c [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+int free_resources()
+{
+    kaa_deinit(kaa_context);
+    return KAA_ERR_NONE;
+}
+
+#endif
+
+KAA_SUITE_MAIN(Protocol_test,initialize,free_resources
```
Comment (ghost) [inline client/client-multi/client-c/test/test_protocol.c]: Please, rename "initialize" with "test_init" and "free_resources" with "test_deinit" to comply with another tests.

File: client/client-multi/client-c/src/kaa/kaa_logging.c [hunk 12]
```diff
@@ -70,6 +70,15 @@ struct kaa_log_collector {
 static const kaa_service_t logging_sync_services[1] = {KAA_SERVICE_LOGGING};
 
 
+kaa_error_t kaa_logging_need_logging_resync(kaa_log_collector_t *self, bool *result)
+{
+    KAA_RETURN_IF_NIL2(self, result, KAA_ERR_BADPARAM);
+    if (self->is_sync_ignored) {
+            return KAA_ERR_NONE;
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_logging.c]: I think the result variable should be initialized explicitly in both cases. Please, also fix code alignment.

File: client/client-multi/client-c/test/test_platform_protocol.c [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    info->allocator_context = mock;
+    mock_strategy_context_t *strategy = (mock_strategy_context_t*) malloc(sizeof(mock_strategy_context_t));
+    mock_storage_context_t *storage = (mock_storage_context_t*) malloc(sizeof(mock_storage_context_t));
+    ASSERT_NOT_NULL(storage);
+    ASSERT_NOT_NULL(strategy);
+    memset(storage, 0, sizeof(mock_storage_context_t));
+    memset(strategy, 0, sizeof(mock_strategy_context_t));
+    kaa_error_t error_code = kaa_logging_init(kaa_context->log_collector, storage, strategy);
+    error_code = kaa_platform_protocol_serialize_client_sync(kaa_context->platfrom_protocol, info, &buffer, &buffer_size);
+    free(info);
```
Comment (ghost) [inline client/client-multi/client-c/test/test_platform_protocol.c]: Use KAA_FREE instead. Please, fix this all around the file.

File: client/client-multi/client-c/test/test_platform_protocol.c [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    info->services_count = 1;
+    info->allocator = &allocator;
+    info->allocator_context = mock;
+    mock_strategy_context_t *strategy = (mock_strategy_context_t*) malloc(sizeof(mock_strategy_context_t));
+    mock_storage_context_t *storage = (mock_storage_context_t*) malloc(sizeof(mock_storage_context_t));
+    ASSERT_NOT_NULL(storage);
+    ASSERT_NOT_NULL(strategy);
+    memset(storage, 0, sizeof(mock_storage_context_t));
+    memset(strategy, 0, sizeof(mock_strategy_context_t));
+    kaa_error_t error_code = kaa_logging_init(kaa_context->log_collector, storage, strategy);
```
Comment (ghost) [inline client/client-multi/client-c/test/test_platform_protocol.c:79]: The result of initialization of the log collector is ignored, it isn't a good practice.

**Comment by ghost** [comment]
LGTM

**Comment by izeigerman** [comment]
LGTM

**Pull request merged**